test-project/src/test_contract_tests.rs

```diff
@@ -127,3 +127,12 @@ async fn test_ink_lang_error() -> Result<()> {
 
     Ok(())
 }
+
+#[tokio::test]
+async fn test_upload() -> Result<()> {
+    let conn = connect_as_test_account().await?;
+
+    assert!(test_contract::upload(&conn).await.is_ok());
```

Comment:
This could return `Ok` if the only thing `upload` did was ... return `Ok` immediately.

Please, add a test that actually checks whether uploaded code:
* did not exist there prior to the first upload
* exists after upload
* can be properly instantiated

Reply:
I so not think is worth it, that I challenge you to write that test (remember to make it idempotent!) if you think it indeed is worth it.

Reply:
I think it is. You should be able to reuse other tests you already wrote - `connect_and_deploy` does most of the things already. There are 4 steps:
1. Connect
2. Deploy
3. Instantiate
4. Call some method

and you already had points 1, 3 and 4 and are now adding 3. Just extract the necessary logic from `connect_and_deploy` and reuse.

Reply:
Not so fast. You cannot `deploy`, because the code is already on the chain. And you cannot remove it from the chain, because there are instances of it.

Reply:
Then this test here has even less sense. If you're testing whether upload works but the contract is already there.